### PR TITLE
[mysql] add custom jdbc properties to debezium mysql connection (#674)

### PR DESCRIPTION
**Pull request by ruanhang1993**
This PR fix the bug that the mysql jdbc connection does not contain the jdbc properties set by users.
The problem is that the debezium mysql connection uses the fixed url pattern.

Changes:
1. add new mysql connection class(MySqlConnectionWithJdbcProperties) and new inner mysql connection configuration class
2. change the usage in mysql cdc from MySqlConnection to MySqlConnectionWithJdbcProperties

### CODE REVIEW

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java [hunk 1]
```diff
@@ -203,7 +203,8 @@ private MySqlBinlogSplit discoverTableSchemasForBinlogSplit(MySqlBinlogSplit spl
         final String splitId = split.splitId();
         if (split.getTableSchemas().isEmpty()) {
             try (MySqlConnection jdbc =
-                    DebeziumUtils.createMySqlConnection(sourceConfig.getDbzConfiguration())) {
+                    DebeziumUtils.createMySqlConnection(
+                            sourceConfig.getDbzConfiguration(), sourceConfig.getJdbcProperties())) {
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java:207]: Why not expose a method `createMySqlConnection(MySqlSourceConfig sourceConfig)` ？

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/connection/MySqlConnectionWithJdbcProperties.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    public String connectionString() {
+        return connectionString(urlPattern);
+    }
+
+    /**
+     * {@link MySqlConnectionConfiguration} extension to be used with {@link
+     * MySqlConnectionWithJdbcProperties}.
+     */
+    public static class MySqlConnectionConfigurationWithCustomUrl
+            extends io.debezium.connector.mysql.MySqlConnection.MySqlConnectionConfiguration {
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/connection/MySqlConnectionWithJdbcProperties.java:52]: Why use full path of `MySqlConnectionConfiguration` here?  We've not another class named `MySqlConnectionConfiguration` , right?

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/connection/MySqlConnectionWithJdbcProperties.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+import io.debezium.config.Configuration;
+import io.debezium.connector.mysql.MySqlConnection;
+import io.debezium.connector.mysql.MySqlConnectorConfig;
+import io.debezium.jdbc.JdbcConnection;
+
+import java.util.Properties;
+
+/** {@link MySqlConnection} extension to be used with MySQL Server. */
+public class MySqlConnectionWithJdbcProperties extends MySqlConnection {
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/connection/MySqlConnectionWithJdbcProperties.java:29]: (1) `MySqlConnectionWithJdbcProperties` can not visit `MySqlConnection`'s default package level methods due to the two classes are lying different package.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/connection/MySqlConnectionWithJdbcProperties.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    private final String urlPattern;
+    /**
+     * Creates a new connection using the supplied configuration.
+     *
+     * @param connectionConfig {@link MySqlConnectionConfiguration} instance, may not be null.
+     */
+    public MySqlConnectionWithJdbcProperties(
+            MySqlConnectionConfigurationWithCustomUrl connectionConfig) {
+        super(connectionConfig);
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/connection/MySqlConnectionWithJdbcProperties.java:38]: (2) We cannot control the parent class how to use `MySqlConnectionConfigurationWithCustomUrl` object in extension way.

As the above (1) and (2), I suggest we directly override `MySqlConnection` under `io.debezium.connector.mysql` package

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/connection/MySqlConnectionWithJdbcProperties.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/connection/MySqlConnectionWithJdbcProperties.java:2]: The license has been outdated

**Review by leonardBang — COMMENTED** [review]
Thanks @ruanhang1993 for the contribution, I left some comments

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/MySqlValidator.java [hunk 6]
```diff
@@ -66,7 +66,9 @@ public void validate() {
                 connection = DebeziumUtils.openJdbcConnection(sourceConfig);
             } else {
                 // for the legacy source
-                connection = DebeziumUtils.createMySqlConnection(from(dbzProperties));
+                connection =
+                        DebeziumUtils.createMySqlConnection(
+                                from(dbzProperties), sourceConfig.getJdbcProperties());
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/MySqlValidator.java:71]:
```suggestion
                        DebeziumUtils.createMySqlConnection(sourceConfig);
```

Comment (ruanhang1993) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/MySqlValidator.java:71]: This is used by the legacy source whose `sourceConfig` is null. I will revert this change to use `new Properties()`.

**Comment by foxmorg** [comment]
Hi there! Trying to update **com.ververica:flink-connector-mysql-cdc** from **2.2.1** to **2.3.0** version and getting such Exception:

```
Caused by: java.lang.NoSuchMethodError: 'void io.debezium.connector.mysql.MySqlConnection$MySqlConnectionConfiguration.<init>(io.debezium.config.Configuration, java.util.Properties)'

    at com.ververica.cdc.connectors.mysql.debezium.DebeziumUtils.createMySqlConnection(DebeziumUtils.java:78)

    at com.ververica.cdc.connectors.mysql.debezium.DebeziumUtils.createMySqlConnection(DebeziumUtils.java:71)

    at com.ververica.cdc.connectors.mysql.source.reader.MySqlSplitReader.checkSplitOrStartNext(MySqlSplitReader.java:138)

    at com.ververica.cdc.connectors.mysql.source.reader.MySqlSplitReader.fetch(MySqlSplitReader.java:71)

    at org.apache.flink.connector.base.source.reader.fetcher.FetchTask.run(FetchTask.java:58)

    at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.runOnce(SplitFetcher.java:142)

    at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.run(SplitFetcher.java:105)

    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)

    at java.base/java.util.concurrent.FutureTask.run(Unknown Source)

    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)

    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)

    ... 1 more
```

As I understand, the reason because I have two dependencies as result:
**debezium-connector-mysql-1.6.4.Final.jar**
**flink-connector-mysql-cdc-2.3.0.jar**

with the same class **io.debezium.connector.mysql.MySqlConnection$MySqlConnectionConfiguration** but first one with only one constructor. So the question is, why was this class renamed in this commit https://github.com/ververica/flink-cdc-connectors/pull/1217/commits/5415bcf0f43db627cd3604697245cc07a630f808 and how may I fix this issue?

**Comment by foxmorg** [comment]
With 2.4.0 the same issue

**Pull request merged**